lib/Helper/HTMLParser/HttpJsonLdParser.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace OCA\Cookbook\Helper\HTMLParser;
+
+use OCA\Cookbook\Exception\HtmlParsingException;
+use OCP\IL10N;
+use OCA\Cookbook\Service\JsonService;
+
+/**
+ * This class is an AbsractHtmlParser that tries to extract a JSON+LD script from the HTML page.
+ * @author Christian Wolf
+ */
+class HttpJsonLdParser extends AbstractHtmlParser {
+	
+	/**
+	 * @var JsonService
+	 */
+	private $jsonService;
+	
+	public function __construct(IL10N $l10n, JsonService $jsonService) {
+		parent::__construct($l10n);
+		
+		$this->jsonService = $jsonService;
+	}
+	
+	public function parse(\DOMDocument $document): array {
+		$xpath = new \DOMXPath($document);
+		
+		$json_ld_elements = $xpath->query("//*[@type='application/ld+json']");
+		
+		foreach ($json_ld_elements as $json_ld_element) {
+			if (!$json_ld_element || !$json_ld_element->nodeValue) {
+				continue;
+			}
+			
+			try {
+				return $this->parseJsonLdElement($json_ld_element);
```

Comment:
This takes the first element it finds and returns it. What if there are multiple valid ones?

Reply:
Are you talking of different Recipe objects within a single page or different schema.org objects?

The `parseJsonLdElement` will throw an Exception if no recipe is found. So, the case of other schema objects can be neglected as far as I see.

Regarding the multi-recipe pages: I have no clue, what we could do to handle this case better. In fact, it is the user's decision what recipe he wants to import. So, he is responsible for checking if the correct recipe was imported.

Reply:
I meant multiple json ld elements. Best case (most effort) scenario would probably be to parse on client device and prompt the user with a selection to choose from. However, since this is really an edge case I vote for ignoring it until there actually is someone with that use case.